### PR TITLE
Various S3 Fixes

### PR DESCRIPTION
**Pull request by ahouene**
This PR contains several elements required for a normal use of the S3 backend:
- Change marker definition method:
  - it (a specifically named S3 object) used to contain the name of the last modified object
  - now the last modified object's ETag and a timestamp are also written
- Fix default endpoint not carrying protocol:
  - endpoints without protocol aren't supported by minio.
- Fix instances of the marker not being updated correctly.

Tests have been updated to reflect these changes.

### CODE REVIEW

File: backends/s3/s3.go [hunk 1]
```diff
@@ -108,37 +108,31 @@ func (b *Backend) List(ctx context.Context, prefix string) (simpleblob.BlobList,
 		return b.doList(ctx, prefix)
 	}
 
+	b.mu.Lock()
```
Comment (wojas) [inline backends/s3/s3.go:111]: You have moved the lock around network calls that can now block other goroutines for several seconds.

File: backends/s3/s3.go [hunk 2]
```diff
@@ -189,9 +183,39 @@ func (b *Backend) Store(ctx context.Context, name string, data []byte) error {
 		return err
 	}
 	if b.opt.UseUpdateMarker {
+		b.mu.Lock()
```
Comment (wojas) [inline backends/s3/s3.go:186]: Again, this locks around a network request, blocking other goroutines for potentially a long time.

Also note that a defer is not scoped to this block. This is fine now in this case, because it occurs at the end of the function, but can cause trouble if another network operation is added at the end of this function later.

**Review by wojas — CHANGES_REQUESTED** [review]

**Comment by wojas** [comment]
Good catch about needing to update the local cache.

> cf. commits for list of changes

Please also summarize the changes in the PR description. That makes it easier to create release notes and later reference the changes in this PR. 

In this case the local cache updating bugfix is important to mention.

**Comment by ahouene** [comment]
>Please also summarize the changes in the PR description

Updated this along with the requested changes!

File: backends/s3/s3.go [hunk 3]
```diff
@@ -109,36 +109,29 @@ func (b *Backend) List(ctx context.Context, prefix string) (simpleblob.BlobList,
...
 		b.lastList = blobs
-		b.mu.Unlock()
-	} else {
-		b.mu.Lock()
-		blobs = b.lastList
+		b.lastTime = time.Now()
 		b.mu.Unlock()
 	}
-	return blobs.WithPrefix(prefix), nil
+	return b.lastList.WithPrefix(prefix), nil
```
Comment (wojas) [inline backends/s3/s3.go:134]: This is unsafe. You need to return `blobs`, because `b.lastList` access requires holding the lock.

File: backends/s3/s3.go [hunk 4]
```diff
@@ -109,36 +109,29 @@ func (b *Backend) List(ctx context.Context, prefix string) (simpleblob.BlobList,
...
 	current := string(data)
 
-	b.mu.Lock()
-	lastMarker := b.lastMarker
-	age := now.Sub(b.lastTime)
-	b.mu.Unlock()
-
-	var blobs simpleblob.BlobList
-	if current != lastMarker || age >= b.opt.UpdateMarkerForceListInterval {
+	if !exists || b.lastMarker == "" || current != b.lastMarker || time.Since(b.lastTime) >= b.opt.UpdateMarkerForceListInterval {
```
Comment (wojas) [inline backends/s3/s3.go:122]: Any access to `b.last*` requires holding the lock, which is one reason the code created new variables in a critical section.

The other reason is readability.

File: backends/s3/s3.go [hunk 5]
```diff
@@ -189,9 +182,38 @@ func (b *Backend) Store(ctx context.Context, name string, data []byte) error {
...
+		wg.Add(1)
+
+		// Update cache
+		go func() {
+			defer wg.Done()
+			// Update size or add to local list if not present
+			l := b.lastList.Len()
+			idx := sort.Search(l, func(i int) bool { return b.lastList[i].Name >= name })
+			blob := simpleblob.Blob{Name: name, Size: int64(len(data))}
+			b.mu.Lock()
```
Comment (wojas) [inline backends/s3/s3.go:195]: Since this is a separate function and the distance between the Lock and Unlock is quite large, I would actually use `defer b.mu.Unlock()` here.

One big advantage of `defer` is that you can be sure it will still be run in case of an early return or panic. There is no such place now, but it's not immediately obvious without inspecting the code between the Lock and Unlock, and someone might inadvertently add one later.

File: backends/s3/s3.go [hunk 6]
```diff
@@ -189,9 +182,38 @@ func (b *Backend) Store(ctx context.Context, name string, data []byte) error {
...
 	}
 	if b.opt.UseUpdateMarker {
+		var wg sync.WaitGroup
+		wg.Add(1)
+
+		// Update cache
+		go func() {
+			defer wg.Done()
+			// Update size or add to local list if not present
+			l := b.lastList.Len()
```
Comment (wojas) [inline backends/s3/s3.go:192]: Accessing `b.lastList` require a lock.

File: backends/s3/s3.go [hunk 7]
```diff
@@ -217,6 +239,33 @@ func (b *Backend) Delete(ctx context.Context, name string) error {
...
 	}
+	if b.opt.UseUpdateMarker {
+		var wg sync.WaitGroup
+		wg.Add(1)
+
+		go func() {
+			defer wg.Done()
+			// Update cached list values
+			// This keeps sorted order, so no need to sort again
+			l := b.lastList.Len()
```
Comment (wojas) [inline backends/s3/s3.go:250]: Again an access without lock.

File: backends/s3/s3.go [hunk 8]
```diff
@@ -189,9 +182,38 @@ func (b *Backend) Store(ctx context.Context, name string, data []byte) error {
...
+			idx := sort.Search(l, func(i int) bool { return b.lastList[i].Name >= name })
+			blob := simpleblob.Blob{Name: name, Size: int64(len(data))}
+			b.mu.Lock()
+			if idx < l {
+				if b.lastList[idx].Name == name {
+					b.lastList[idx].Size = int64(len(data))
+				} else {
+					b.lastList = append(b.lastList, simpleblob.Blob{})
+					copy(b.lastList[idx+1:], b.lastList[idx:])
```
Comment (wojas) [inline backends/s3/s3.go:201]: The whole clever insert causes race conditions with locking.

If you always set the attributes to a completely new value, it is safe to return that value and use access it when not locked, which is what we do.

With the insertion trick, the values are no longer immutable and can no longer be accessed without holding a lock.

In the old version `b.lastList` could safely be returned to the caller and the caller could access the items without holding a lock.

In the new version accessing the items in the slice without holding the lock (an implementation detail) can/will cause race conditions when the list is updated at the same time.
 
Instead of the insert, create a new slice and then assign the new slide to `b.lastList` in a tiny critical section. Something like:

```go
b.mu.Lock()
oldList := b.lastList 
b.mu.Unlock()

var newList []simpleblob.Blob
// fill newList here

b.mu.Lock()
b.lastList = newList
b.mu.Unlock()
```

**Review by wojas — CHANGES_REQUESTED** [review]

**Comment by ahouene** [comment]
Going for cache invalidation instead of local cache update, at least for now.

**Comment by wojas** [comment]
The description could use a few words on how this PR affects the update marker optimisation and why. Even after reading the changes twice it is not clear to me what it fixes or improves and if the changes are correct.

File: backends/s3/s3.go [hunk 9]
```diff
@@ -109,35 +109,35 @@ func (b *Backend) List(ctx context.Context, prefix string) (simpleblob.BlobList,
...
 	}
 	current := string(data)
 
 	b.mu.Lock()
-	lastMarker := b.lastMarker
-	age := now.Sub(b.lastTime)
+	shouldUpdate := b.lastList == nil || b.lastMarker != current ||
+		time.Since(b.lastTime) >= b.opt.UpdateMarkerForceListInterval ||
+		!exists
+	blobs := b.lastList
```
Comment (wojas) [inline backends/s3/s3.go:126]: Good change, this guarantees that we take a reference to the blobs at the time we checked the other state.

It probably would not have mattered that we previously could have returned a newer version if they got updated in a different goroutine, but this is cleaner.

File: backends/s3/s3.go [hunk 10]
```diff
@@ -217,6 +222,15 @@ func (b *Backend) Delete(ctx context.Context, name string) error {
 	if err = handleErrorResponse(err); err != nil {
 		metricCallErrors.WithLabelValues("delete").Inc()
 	}
+	if b.opt.UseUpdateMarker {
+		if err := b.doStore(ctx, UpdateMarkerFilename, []byte(name)); err != nil {
```
Comment (wojas) [inline backends/s3/s3.go:226]: This introduces buggy behaviour.

Previously the update marker would contain the filename of any newly added file, signaling that something changed. These file are assumed to be unique between instances (no two instances will try to write the same name). If this assumption is not correct, the update marker optimisation MUST NOT be used.

But now with this change you are also setting this to indicate that you deleted the file. Consider the following sequence of events:

- Instance A writes foo.txt and updates the marker to 'foo.txt'
- Instance D lists the files and records the last marker as 'foo.txt'
- Instance B writes bar.txt and updates the marker to 'bar.txt'
- Instance C removes 'foo.txt' and updates the marker to 'foo.txt'
- Instance D lists the files again and determines that nothing has changed, because the marker did not change.

Oops.

If you want to update the marker in this case, you have to set it to a different value.

Perhaps we can set the marker to a combination of:

- the blob name affected
- the nanosecond timestamp of the change (or a hash of the contents)

This would allow the mechanism to also work correctly for both updates and deletes. For delete you could also add an indicator that the change was a deletion.

A hash would provide slightly stronger guarantees than a timestamp, but it adds overhead (especially for bigger blobs) and we already have the forced periodic reload to handle these.

File: backends/s3/s3_test.go [hunk 11]
```diff
@@ -76,7 +77,7 @@ func getBackend(ctx context.Context, t *testing.T) (b *Backend) {
 
 func TestBackend(t *testing.T) {
 	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
-	t.Cleanup(cancel)
```
Comment (wojas) [inline backends/s3/s3_test.go:79]: Just curious, why this change?

File: test.sh [hunk 12]
```diff
@@ -3,33 +3,41 @@
 set -e
 
 # Stop all child processes on exit
-trap "pkill -P $$" EXIT
+trap cleanup EXIT
+
+cleanup() {
+    pkill -P $$
+    if [ -d "$tmpdir" ]; then
```
Comment (wojas) [inline test.sh:10]: I would feel better if this checked if it explicitly checked if `$tmpdir` was not empty, even though bash does appear to do the right thing (`-d` on an empty string returns false). 

```suggestion
    if [ ! -z "$tmpdir" ] && [ -d "$tmpdir" ]; then
```

I am very paranoid about commands that recursively delete directories.

**Review by wojas — CHANGES_REQUESTED** [review]

File: backends/s3/s3.go (hunk 10 above)
Comment (ahouene) [reply backends/s3/s3.go:226]: Thank you!
I'll probably use a combination of the blob name, the operation type (store/delete) and the nanosecond timestamp. Then comparing the ETag should be enough, wouldn't it?
Also, as using the hash of the content would be safer, maybe replacing the timestamp with the ETag of the blob that was just uploaded makes sense?

File: backends/s3/s3.go [hunk 13]
```diff
@@ -189,9 +189,14 @@ func (b *Backend) Store(ctx context.Context, name string, data []byte) error {
 		return err
 	}
 	if b.opt.UseUpdateMarker {
+    		b.lastMarker = name
```
[the remark this answers is not in the document]
Comment (ahouene) [reply backends/s3/s3.go:192]: That was a mistake, sorry!

File: backends/s3/s3_test.go (hunk 11 above)
Comment (ahouene) [reply backends/s3/s3_test.go:79]: Hmmm, probably a misunderstanding of the role of `t.Cleanup` at the time I wrote this?

File: test.sh (hunk 12 above)
Comment (ahouene) [reply test.sh:10]: Better paranoid than careless! I'll add this.

File: backends/s3/s3.go (hunk 10 above)
Comment (wojas) [reply backends/s3/s3.go:226]: An ETag would be fine too, but I don't know if we can rely on every S3 API implementation returning one. Perhaps use it when available, and fall back to a timestamp if not?

File: backends/s3/marker.go [hunk 14]
```diff
@@ -13,12 +13,15 @@ import (
...
+func (b *Backend) setMarker(ctx context.Context, name, etag string, isDel bool) error {
 	if !b.opt.UseUpdateMarker {
 		return nil
 	}
 	ts := strconv.FormatInt(time.Now().UnixNano(), 10)
 	lines := []string{name, etag, ts}
+	if isDel {
+    		lines = append(lines, "del")
+	}
 	joined := []byte(strings.Join(lines, "\x00"))
```
Comment (wojas) [inline backends/s3/marker.go:25]: Nit: using `fmt.Sprintf` is cleaner and probably faster (fewer allocs).

Comment (ahouene) [reply backends/s3/marker.go]: You're right. Just pushed fix.

File: backends/s3/marker.go [hunk 15]
```diff
@@ -17,19 +16,16 @@ func (b *Backend) setMarker(ctx context.Context, name, etag string, isDel bool)
...
 	}
-	ts := strconv.FormatInt(time.Now().UnixNano(), 10)
-	lines := []string{name, etag, ts}
-	if isDel {
-    		lines = append(lines, "del")
-	}
-	joined := []byte(strings.Join(lines, "\x00"))
-	_, err := b.doStore(ctx, UpdateMarkerFilename, joined)
+	nanos := time.Now().UnixNano()
+	s := fmt.Sprintf("%s\x00%s\x00%d\x00%v", name, etag, nanos, isDel)
```
Comment (wojas) [inline backends/s3/marker.go:20]: I would prefer a readable string as marker, all the fields in here are readable strings.

```suggestion
	s := fmt.Sprintf("%s:%s:%d:%v", name, etag, nanos, isDel)
```

File: test.sh [hunk 16]
```diff
@@ -3,33 +3,41 @@
...
+        chmod u+x "$dst"
     fi
 
     # Start MinIO
-    echo "* Starting $minio on port 34730"
-    tmpdir=$(mktemp -d -t minio.XXXXXX)
-    "$minio" server --address 127.0.0.1:34730 --console-address 127.0.0.1:34731 --quiet "$tmpdir" &
-    sleep 3
+    echo "* Starting minio at address 127.0.0.1:34730"
+    tmpdir=$(mktemp --directory --tmpdir=. .minio.XXXXXX)
```
Comment (wojas) [inline test.sh:35]: This fails on macOS:

```
mktemp: illegal option -- -
usage: mktemp [-d] [-q] [-t prefix] [-u] template ...
       mktemp [-d] [-q] [-u] -t prefix
```

File: backends/s3/s3.go [hunk 17]
```diff
@@ -103,7 +104,7 @@ type Backend struct {
 	log    logr.Logger
 
 	mu         sync.Mutex
-	lastMarker string
+	lastMarker []byte
```
Comment (wojas) [inline backends/s3/s3.go:107]: Making this a `[]byte` makes it harder to use and there is no reason this should not be a readable value.

File: backends/s3/s3.go [hunk 18]
```diff
@@ -169,57 +169,64 @@ func (b *Backend) doList(ctx context.Context, prefix string) (simpleblob.BlobLis
...
 
-	_, err := b.client.PutObject(ctx, b.opt.Bucket, name, bytes.NewReader(data), int64(len(data)), minio.PutObjectOptions{
+	info, err := b.client.PutObject(ctx, b.opt.Bucket, name, bytes.NewReader(data), int64(len(data)), minio.PutObjectOptions{
 		NumThreads: 3,
 	})
 	if err != nil {
 		metricCallErrors.WithLabelValues("store").Inc()
 	}
-	return err
+	return &info, err
```
Comment (wojas) [inline backends/s3/s3.go:212]: If you need to take the address of this to convert it into the `*minio.UploadInfo`, you probably should just return `minio.UploadInfo` instead.

File: backends/s3/s3.go [hunk 19]
```diff
@@ -288,8 +295,10 @@ func New(ctx context.Context, opt Options) (*Backend, error) {
 		// Ok, no SSL
 	case "https":
 		useSSL = true
+	case "":
+		return nil, fmt.Errorf("No scheme provided for endpoint URL '%s', use http or https.", opt.EndpointURL)
 	default:
-		return nil, fmt.Errorf("unsupported scheme for S3: '%s'", u.Scheme)
+		return nil, fmt.Errorf("Unsupported scheme for S3: '%s', use http or https.", u.Scheme)
```
Comment (wojas) [inline backends/s3/s3.go:301]: Nit: errors should start with lower case chars.

File: test.sh [hunk 20]
```diff
@@ -3,33 +3,41 @@
...
+    minio server --address 127.0.0.1:34730 --console-address 127.0.0.1:34731 --quiet "$tmpdir" &
+    # Wait for minio server to be ready
+    while ! curl -s -I "127.0.0.1:34730/minio/health/ready" | grep '200 OK' >/dev/null; do
```
Comment (wojas) [inline test.sh:38]: This will hang forever if there is something wrong with minio or how you call it.

File: test.sh [hunk 21]
```diff
@@ -39,6 +47,7 @@ set -ex
 go test -count=1 "$@" ./...
 
 # Configure linters in .golangci.yml
-go install github.com/golangci/golangci-lint/cmd/golangci-lint@v1.50.1
-./bin/golangci-lint run
-
+if ! command -v golangci-lint >/dev/null; then
```
Comment (wojas) [inline test.sh:50]: Not good enough, we need that specific version, which is why we go install the specific version. Older versions can fail on newer Go releases and vice versa.

File: backends/s3/s3_test.go [hunk 22]
```diff
@@ -74,24 +78,27 @@ func getBackend(ctx context.Context, t *testing.T) (b *Backend) {
...
 
-	data, err := b.Load(ctx, UpdateMarkerFilename)
+	markerFileContent, err := b.Load(ctx, UpdateMarkerFilename)
+	assert.NoError(t, err)
+	assert.True(t, bytes.HasPrefix(markerFileContent, []byte("foo-1")))
+
+	markerFileContent, err = b.Load(ctx, UpdateMarkerFilename)
 	assert.NoError(t, err)
-	assert.Equal(t, "bar-1", string(data))
+	assert.Equal(t, b.lastMarker, markerFileContent)
```
Comment (wojas) [inline backends/s3/s3_test.go:103]: By removing the tests for expected content, this test no longer tests the marker. it could be an empty string here and the test would be totally happy.

Since this test is here to test the marker, test for the specific contents you expect. You cannot control the timestamp and external ETag, but you can check if it starts with the filename.

File: backends/s3/s3_test.go [hunk 23]
```diff
@@ -74,24 +78,27 @@ func getBackend(ctx context.Context, t *testing.T) (b *Backend) {
...
 func TestBackend_marker(t *testing.T) {
 	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
-	t.Cleanup(cancel)
+	defer cancel()
 
 	b := getBackend(ctx, t)
 	b.opt.UseUpdateMarker = true
 
 	tester.DoBackendTests(t, b)
-	assert.Equal(t, "bar-1", b.lastMarker)
```
Comment (wojas) [inline backends/s3/s3_test.go:92]: Same here, this test needs to remain, but be updated for the new format.

File: backends/s3/s3_test.go (hunk 22 above)
Comment (wojas) [reply backends/s3/s3_test.go:103]: Ah wait, you are testing the prefix above. Then this is fine, but you still need to add back the "bar-1" test.

File: backends/s3/s3_test.go [hunk 24]
```diff
@@ -51,6 +52,9 @@ func getBackend(ctx context.Context, t *testing.T) (b *Backend) {
 	require.NoError(t, err)
 
 	cleanup := func() {
+		ctx, cancel := context.WithCancel(context.Background())
```
[the remark this answers is not in the document]
Comment (ahouene) [reply backends/s3/s3_test.go:55]: My thought here was that the outer-scope `ctx` might have been cancelled by the time `cleanup` is called. Should I try and use the outer-scope one regardless? Or make my intent clearer with something like `if ctx.Err() != nil { ctx = ... }`?

File: backends/s3/s3_test.go (hunk 23 above)
Comment (ahouene) [reply backends/s3/s3_test.go:92]: I updated it!

File: backends/s3/s3_test.go (hunk 22 above)
Comment (ahouene) [reply backends/s3/s3_test.go]: I changed the prefix test in favour of a regexp test.

Not sure if you mean "bar-1", the value, or if you mean the test. In case you mean the value "bar-1" needs to appear, it is no longer true because the last write operation on the backend (in DoBackendTests) is the deletion of the blob named "foo-1".

File: test.sh (hunk 21 above)
Comment (ahouene) [reply test.sh]: Oh, indeed `golangci-lint` can be moody!
I now check the output of `golangci-lint version`.

File: test.sh (hunk 20 above)
Comment (ahouene) [reply test.sh:39]: Now capped at 100 attempts (10s if we ignore the call time of `curl`)

File: test.sh (hunk 16 above)
Comment (ahouene) [reply test.sh]: Hah, I should be more POSIX-y. Good catch, that would have failed on BSD and BusyBox too!

File: backends/s3/s3_test.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (wojas) [reply backends/s3/s3_test.go:54]: That outer scope will indeed likely be cancelled by the time this runs, because of the defer. The t.Cleanup that was called earlier would have been run after this cleanup function, so things would probably have worked as expected. But since you changed the t.Cleanup into a defer, the context will be closed before this is run.

File: test.sh [hunk 25]
```diff
@@ -3,33 +3,49 @@
...
+    i=0
+    while ! curl -s -I "127.0.0.1:34730/minio/health/ready" | grep '200 OK' >/dev/null; do
+        i=$((i+1))
+        if [ "$i" -ge 100 ]; then
+            # We have been waiting for server to start for 10 seconds
+            echo "Minio could not start properly"
+            curl -s -I "127.0.0.1:34730/minio/health/ready"
+            exit 1
+        fi
+        sleep .1
```
Comment (wojas) [inline test.sh:47]: From the macOS manpage:

> The sleep command allows and honors a non-integer number of seconds to sleep in any form acceptable by strtod(3).  This is a non-portable extension, but is also implemented in GNU sh-utils since version 2.0a (released in 2002).

I think this is ok for now, but it may not work on every platform.

File: test.sh [hunk 26]
```diff
@@ -39,6 +55,8 @@ set -ex
 go test -count=1 "$@" ./...
 
 # Configure linters in .golangci.yml
-go install github.com/golangci/golangci-lint/cmd/golangci-lint@v1.50.1
-./bin/golangci-lint run
-
+expected_version=v1.50.1
+if ! golangci-lint version | grep -w "$expected_version" >/dev/null 2>&1; then
```
Comment (wojas) [inline test.sh:59]: No need for all this complexity. If the right version is already installed, `go install` is very fast.

File: test.sh (hunk 21 above)
Comment (wojas) [reply test.sh]: No need for all this complexity. If the right version is already installed, go install is very fast.

File: backends/s3/s3_test.go (hunk 22 above)
Comment (wojas) [reply backends/s3/s3_test.go]: There was one test above that you removed.

**Review by wojas — APPROVED** [review]
